cpp/include/cudf/hashing/detail/murmurhash3_x86_32.cuh

```diff
@@ -24,104 +24,169 @@
 #include <cudf/structs/struct_view.hpp>
 #include <cudf/types.hpp>
 
-#include <cuco/hash_functions.cuh>
-#include <cuda/std/cstddef>
+#include <cstddef>
 
 namespace cudf::hashing::detail {
 
+// MurmurHash3_x86_32 implementation from
+// https://github.com/aappleby/smhasher/blob/master/src/MurmurHash3.cpp
+//-----------------------------------------------------------------------------
+// MurmurHash3 was written by Austin Appleby, and is placed in the public
+// domain. The author hereby disclaims copyright to this source code.
+// Note - The x86 and x64 versions do _not_ produce the same results, as the
+// algorithms are optimized for their respective platforms. You can still
+// compile and run any of them on any platform, but your performance with the
+// non-native version will be less than optimal.
 template <typename Key>
 struct MurmurHash3_x86_32 {
   using result_type = hash_value_type;
 
-  CUDF_HOST_DEVICE constexpr MurmurHash3_x86_32(uint32_t seed = cudf::DEFAULT_HASH_SEED)
-    : _impl{seed}
+  constexpr MurmurHash3_x86_32() = default;
+  constexpr MurmurHash3_x86_32(uint32_t seed) : m_seed(seed) {}
+
+  [[nodiscard]] __device__ inline uint32_t fmix32(uint32_t h) const
   {
+    h ^= h >> 16;
+    h *= 0x85ebca6b;
+    h ^= h >> 13;
+    h *= 0xc2b2ae35;
+    h ^= h >> 16;
+    return h;
   }
 
-  __device__ constexpr result_type operator()(Key const& key) const { return this->_impl(key); }
+  [[nodiscard]] __device__ inline uint32_t getblock32(std::byte const* data,
+                                                      cudf::size_type offset) const
+  {
+    // Read a 4-byte value from the data pointer as individual bytes for safe
+    // unaligned access (very likely for string types).
+    auto const block = reinterpret_cast<uint8_t const*>(data + offset);
+    return block[0] | (block[1] << 8) | (block[2] << 16) | (block[3] << 24);
+  }
 
-  __device__ constexpr result_type compute_bytes(cuda::std::byte const* bytes,
-                                                 std::uint64_t size) const
+  [[nodiscard]] result_type __device__ inline operator()(Key const& key) const
   {
-    return this->_impl.compute_hash(bytes, size);
+    return compute(normalize_nans_and_zeros(key));
   }
 
- private:
   template <typename T>
-  __device__ constexpr result_type compute(T const& key) const
+  result_type __device__ inline compute(T const& key) const
   {
-    return this->compute_bytes(reinterpret_cast<cuda::std::byte const*>(&key), sizeof(T));
+    return compute_bytes(reinterpret_cast<std::byte const*>(&key), sizeof(T));
```

Comment:
Note that if you pass `sizeof(T)` as a plain `int`, you likely end up storing it into a register, even though its compile time value. You can avoid this by passing it as `cuco::extent<size_t, sizeof(T)>{}`.